include/zephyr/usb/class/usbd_hid.h

```diff
@@ -0,0 +1,207 @@
+	 * configuration and the application can e.g. begin submitting input
+	 * reports, and with the argument set to false when the interface is no
+	 * longer active. This callback is optional.
+	 */
+	void (*iface_ready)(const struct device *dev, const bool ready);
+
+	/**
+	 * This callback is called for the HID Get Report request to get a
+	 * feature, input, or output report, which is specified by the argument
+	 * type. If there is no report ID in the report descriptor, the id
+	 * argument is zero. The callback implementation must check the
+	 * arguments, such as whether the report type is supported, and return
+	 * a nonzero value to indicate an unsupported type or an error.
+	 */
+	int (*get_report)(const struct device *dev,
+			  const uint8_t type, const uint8_t id,
+			  const uint16_t len, uint8_t *const buf);
+
+	/**
+	 * This callback is called for the HID Set Report request to set a
+	 * feature, input, or output report, which is specified by the argument
+	 * type. If there is no report ID in the report descriptor, the id
+	 * argument is zero. The callback implementation must check the
+	 * arguments, such as whether the report type is supported, and return
+	 * a nonzero value to indicate an unsupported type or an error.
+	 */
+	int (*set_report)(const struct device *dev,
+			  const uint8_t type, const uint8_t id,
+			  const uint16_t len, const uint8_t *const buf);
+
+	/**
+	 * Notification to limit intput report frequency.
+	 * The application should mute an input report submission until a new
+	 * event occurs or until the time specified by the duration value has
+	 * elapsed. If a report ID is used in the report descriptor, the
+	 * application must store the duration and handle the specified report
+	 * accordingly.
+	 */
+	void (*set_idle)(const struct device *dev,
+			 const uint8_t id, const uint32_t duration);
+
+	/**
+	 * If a report ID is used in the report descriptor, the application
+	 * must implement this callback and return the duration for the
+	 * specified report ID.
+	 */
+	uint32_t (*get_idle)(const struct device *dev, const uint8_t id);
+
+	/**
+	 * Notification that the host has changed the protocol from
+	 * Boot Protocol(0) to Report Protocol(1) or vice versa.
+	 */
+	void (*set_protocol)(const struct device *dev, const uint8_t proto);
+
+	/**
+	 * Notification that input report submitted with
+	 * hid_device_submit_report() has been sent.
+	 */
+	void (*input_report_done)(const struct device *dev);
+
+	/**
+	 * New output report callback. Callback will only be called for reports
+	 * received through the optional interrupt OUT pipe. If there is no
+	 * interrupt OUT pipe, output reports will be received using set_report().
+	 * If a report ID is used in the report descriptor, the host places the ID
+	 * in the buffer first, followed by the report data.
+	 */
+	void (*output_report)(const struct device *dev, const uint16_t len,
+			      const uint8_t *const buf);
```

Comment:
Why not feed the data received through the OUT interrupt pipe also through the set_report callback instead? The upper level HID application should be agnostic to the detail of which pipe the output report arrived on, especially if the intention is to make this API transport independent.

Reply:
I had changed it to use set_report, not sure if the commit was published in that PR, but changed it back because I found it confusing. In the Get/Set Report, there is an explicit report id parameter. Report id is optional, but then the id parameter is reserved 0 and everything is clear. For the input report, the IN endpoint should be used, although the input report can also be retrieved using Get Report, the presentation is slightly different. So different representations -> different interfaces. For output reports, I find it more intuitive to make them similar to input report interface and not abuse Set Report with ID 0. So it can be useful to know which interface is being used, passing it through the same callback as Set Report will remove this information. The application, like keyboard sample, can just call the same input report code with ID 0 if there are no IDs in the report descriptor.

Reply:
Well if there is a report ID parameter, that report ID will also occupy the first byte of the transferred report, so passing the report ID as an extra argument in the set_report is redundant. The same high level application might be ported to different targets with or without a dedicated OUT endpoint, which might create some inconvenience.

How about keeping the separate callback, but if it's not implemented by the application, the driver falls back to calling the set_report?

Reply:
> How about keeping the separate callback, but if it's not implemented by the application, the driver falls back to calling the set_report?

I am very much against it. Having or not having an optional OUT endpoint is not a bagatelle. The HID device implementor needs to be aware of the differences. If there is no need for low latency output reports then OUT endpoint should not be used. Also, something like "HID RAW" for regular data transfer should not be used and is discouraged. Btw, ID parameter in set_report is not redundant, interface is different, type and id in set_report are from setup packet, report "payload" from the data stage.

Reply:
> I am very much against it. Having or not having an optional OUT endpoint is not a bagatelle. The HID device implementor needs to be aware of the differences. If there is no need for low latency output reports then OUT endpoint should not be used.

I agree with @jfischer-no , it should be explicit, not a fallback. 

I can see that @benedekkupper would like a more polished, higher-lever HID experience, probably thinking about HID over GATT + HID over USB. That's great as it's own thing, but for the USB, it needs to stick to the way USB HID works, or it will cause a lot of headaches.

Reply:
> Btw, ID parameter in set_report is not redundant, interface is different, type and id in set_report are from setup packet, report "payload" from the data stage.

They are redundant in the sense that the spec requires them to be the same, even if the same values appears twice on the bus.

> Also, something like "HID RAW" for regular data transfer should not be used and is discouraged.

How do you define "regular data transfer"? HID has seen a lot of use as a general purpose middleware transport, just on top of my head there's FIDO, CMSIS DAP and NXP's KBOOT flashing protocols.

In the end I'm ok with the current situation, I understand the justifications now.

Reply:
> How do you define "regular data transfer"? HID has seen a lot of use as a general purpose middleware transport, just on top of my head there's FIDO, CMSIS DAP and NXP's KBOOT flashing protocols.

By regular, I mean something that uses reports as a data payload and does not require periodic transfers. CMSIS-DAP moved to bulk transfers, e.g. discussion in https://github.com/zephyrproject-rtos/zephyr/pull/53798